pennylane/transforms/optimization/merge_rotations.py

```diff
@@ -187,48 +186,36 @@ def stop_at(obj):
+                next_params = qml.math.stack(next_gate.parameters, like=interface)
+                # jax-jit does not support cast_like
+                if not qml.math.is_abstract(cumulative_angles):
+                    next_params = qml.math.cast_like(next_params, cumulative_angles)
+
                 # The Rot gate must be treated separately
                 if current_gate.name == "Rot":
-                    if is_abstract(cumulative_angles):
-                        # jax-jit does not support cast_like
-                        cumulative_angles = cumulative_angles + stack(next_gate.parameters)
-                    else:
-                        cumulative_angles = fuse_rot_angles(
-                            cumulative_angles,
-                            cast_like(
-                                stack(next_gate.parameters, like=interface), cumulative_angles
-                            ),
-                        )
+                    cumulative_angles = fuse_rot_angles(cumulative_angles, next_params)
                 # Other, single-parameter rotation gates just have the angle summed
                 else:
-                    if is_abstract(cumulative_angles):
-                        # jax-jit does not support cast_like
-                        cumulative_angles = cumulative_angles + stack(next_gate.parameters)
-                    else:
-                        cumulative_angles = cumulative_angles + cast_like(
-                            stack(next_gate.parameters, like=interface), cumulative_angles
-                        )
+                    cumulative_angles = cumulative_angles + next_params
             # If it is not, we need to stop
             else:
                 break
 
             # If we did merge, look now at the next gate
             next_gate_idx = find_next_gate(current_gate.wires, list_copy[1:])
 
-        # If we are tracing/jitting, don't perform any conditional checks and
+        # If we are tracing/jitting or differentiating, don't perform any conditional checks and
         # apply the operation regardless of the angles. Otherwise, only apply if
         # the rotation angle is non-trivial.
-        if is_abstract(cumulative_angles):
+        if (
+            qml.math.is_abstract(cumulative_angles)
+            or qml.math.requires_grad(cumulative_angles)
+            or not qml.math.allclose(cumulative_angles, 0.0, atol=atol, rtol=0)
+        ):
             with QueuingManager.stop_recording():
                 new_operations.append(
                     current_gate.__class__(*cumulative_angles, wires=current_gate.wires)
                 )
-        else:
-            if not allclose(cumulative_angles, zeros(len(cumulative_angles)), atol=atol, rtol=0):
-                with QueuingManager.stop_recording():
-                    new_operations.append(
-                        current_gate.__class__(*cumulative_angles, wires=current_gate.wires)
-                    )
```

Comment:
Only use the optimization of skipping gates if the angles were not differentiable (because otherwise the derivative is altered by skipping the gate).
Also, condense the code into one if-clause to avoid code repetition.